### PR TITLE
Add a "Past-due" view for scheduled actions

### LINKED ISSUE #510: Add a "Past" view for pending scheduled actions
Add a view that shows only actions that are pending and have past their scheduled date. This would be especially helpful for the message shown in #91.

Suggested by @rrennick in Slack DM.

### PR DESCRIPTION
**Pull request by crstauf**
Closes #510.

- [ ] After #509 and this PR have been merged, add a link to the view on the past-due actions admin notice (intro'd in #509).

### Changelog

> Enhancement - Adds a new "Past Due" view to the scheduled actions list table.

### CODE REVIEW

**Comment by crstauf** [comment]
@rrennick Job #1772.21 failed because of a unix timestamp comparison failure; can you please trigger a rebuild?

**Comment by rrennick** [comment]
@crstauf triggered & passed.

**Comment by crstauf** [comment]
@rrennick How can the PR alert custom data stores to support the "Past-due" view?

In the PR, I've adjusted the built-in data stores, which means that custom data stores will not have the necessary code to support the "Past-due" view.

Suggestions?

**Comment by crstauf** [comment]
@rrennick @thenbrent This PR is ready for review.

**Comment by rrennick** [comment]
@crstauf Thanks again. On initial look on this looks good. I'll hold it until the past due notice is merged.

**Comment by jorgeatorres** [comment]
Re-assigning to @barryhughes per p1659719396500269-slack-C0E1AV8T0.

File: classes/abstracts/ActionScheduler_Store.php [hunk 1]
```diff
@@ -131,6 +131,28 @@ public function query_action( $query ) {
...
+		$pastdue_action_counts = ( int ) $this->query_actions( array(
+			'status' => self::STATUS_PENDING,
+			'date'   => as_get_datetime_object(),
+		), 'count' );
+
+		if ( $pastdue_action_counts ) {
+			$extra_actions['past-due'] = $pastdue_action_counts;
+		}
+
+		return apply_filters( 'action_scheduler_extra_action_counts', $extra_actions );
```
Comment (barryhughes) [inline classes/abstracts/ActionScheduler_Store.php:159]: Add a docblock?

**Review by barryhughes — COMMENTED** [review]
Observation: as currently defined, past-due actions also include *async* actions, and because async actions have a nominal scheduled date of something like `0000-00-00 00:00:00` they are also listed first (and while it's true they might legitimately be past-due, sometimes nothing will actually be wrong and they will simply be waiting on the next queue runner starting).

https://user-images.githubusercontent.com/3594411/183159488-633cc8d2-74a1-4f19-9529-500d11b90746.mov

☝🏽 This probably won't be too much of a problem, most of the time, but I wanted to surface it. It also means that occasionally the **Past-Due** view filter will be displayed when actually all we have are some async actions that are nothing to worry about:

past-due-view

Lastly, and though I know it's not technically new as of this PR, the past-due admin notice may appear for similar reasons (even if there are no past-due *scheduled* actions):

warning-past-due

Again, I realize some of the code pre-dates this PR, but I wanted to get some thoughts before merging.

File: classes/abstracts/ActionScheduler_Store.php (hunk 1 above)
Comment (jorgeatorres) [reply classes/abstracts/ActionScheduler_Store.php:159]: Done.

**Comment by jorgeatorres** [comment]
Hi @barryhughes!

Thanks for the review! I've added the docblock per your request.

As for the async actions, I totally get what you mean. For those, "past due" means something different, but I'm not sure if there's an easy way to get the correct count using the query args that the AS datastore currently supports.
I guess we could get "creative": for instance, we could ignore async actions when checking whether there are any actions past due (maybe by subtracting actions with a scheduled date <= `0000-00-00 ...`). Would that make sense?

That's not ideal as async actions can be "past due" in a way that makes sense in this check but we don't have the information nor query args to support that so we could at least try to reduce noise.

Please let me know if you think of a better way to do this check or if I'm missing something. Thanks!

**Comment by barryhughes** [comment]
Yep, I think you're right: unless we also want to make a schema change, there isn't really a clean and efficient way to handle this. I think we could ship this as it is (and we can make further enhancements in the future if feedback suggests it is needed). wdyt?

**Review by barryhughes — APPROVED** [review]

**Comment by jorgeatorres** [comment]
>  I think we could ship this as it is (and we can make further enhancements in the future if feedback suggests it is needed). wdyt?

Sounds good to me. If necessary, we could then do something along the lines I described above or revisit [this issue](https://github.com/woocommerce/action-scheduler/issues/686) on adding a way to search for async actions.

**Pull request merged**